crates/motionfile/src/motion_interpolator.rs

```diff
@@ -1,157 +1,213 @@
-            match item {
-                MotionItem::Spline(spline) => spline.reset(),
-                MotionItem::Condition(condition) => condition.reset(),
-            }
+        self.current_state = State::CheckEntry {
+            current_index: 0,
+            time_since_start: Duration::ZERO,
         }
-        self.index = 0;
     }
 
-    fn update(&mut self, sensor_data: &SensorData) {
-        self.items.iter_mut().for_each(|item| {
-            if let MotionItem::Condition(condition) = item {
-                condition.update(sensor_data)
-            }
-        })
+    pub fn current_time(&self) -> Duration {
+        match self.current_state {
+            State::CheckEntry { time_since_start, .. } => time_since_start,
+            State::InterpolateSpline { time_since_start, .. } => time_since_start,
+            State::CheckExit { time_since_start, .. } => time_since_start,
+            State::Finished => Duration::ZERO,
+        }
     }
 }
 
 impl<T: Debug + Interpolate<f32>> TryFrom<MotionFile<T>> for MotionInterpolator<T> {
     type Error = Report;
 
     fn try_from(motion_file: MotionFile<T>) -> Result<Self> {
-        let mut current_time = Duration::ZERO;
-        let mut current_spline_frames = vec![Key::new(
-            current_time,
-            motion_file.initial_positions,
-            Interpolation::Linear,
-        )];
-
-        let mut motion_items = Vec::new();
-
-        for frame in motion_file.frames {
-            match frame {
-                MotionFileFrame::Joints {
-                    duration,
-                    positions,
-                } => {
-                    current_time += duration;
-                    current_spline_frames.push(Key::new(
-                        current_time,
-                        positions,
-                        Interpolation::Linear,
-                    ));
-                }
-                MotionFileFrame::Condition(condition) => {
-                    motion_items.push(MotionItem::Spline(SplineInterpolator::try_new(
-                        current_spline_frames.clone(),
-                    )?));
-                    let last = current_spline_frames.pop().unwrap();
-                    current_spline_frames.clear();
-                    current_spline_frames.push(last);
-                    motion_items.push(MotionItem::Condition(condition));
-                }
-            }
-        }
-        if current_spline_frames.len() > 1 {
-            motion_items.push(MotionItem::Spline(SplineInterpolator::try_new(
-                current_spline_frames,
-            )?));
-        }
+        let first_frame = motion_file.motion.first().unwrap();
+        let mut first_frame_keyframes: VecDeque<_> = first_frame.keyframes.clone().into();
+        let initial_positions = first_frame_keyframes.pop_front().unwrap().positions;
+
+        let mut motion_frames = vec![ConditionedSpline {
+            entry_condition: first_frame.entry_condition.clone(),
```

Comment:
yes, since we cannot move out of the vec because we iterate over all items later. However this clone happens only once in initialization and is relatively cheap

Reply:
Okay